### ISSUE TITLE
Breaking change in `load_model` from `1.14.1` -> `1.15.0`

### ISSUE THREAD

**Issue by tgolsson** [onnx/onnx#5746]
# Bug Report

### Describe the bug

This isn't *really* a bug per se, more of a SemVer violation, and I figured I'd file it because it hit us and someone else might search for it. I don't think it's actionable at this point, because it's likely to affect a small amount of users, and working around it would likely be overly permissive instead.

Either way, the break happened as follows. In 1.14.1, `load_model` was defined like this:
```py3
def load_model(
    f: Union[IO[bytes], str],
    format: Optional[Any] = None,  # pylint: disable=redefined-builtin
    load_external_data: bool = True,
) -> ModelProto:
```

This API says that we can optionally provide a format of any type. Due to perhaps misunderstanding the parameter, our code ended up calling it this way:
```py3
                model_proto = onnx.load_model(f, onnx.ModelProto)
```
Which is a bit nonsensical on the one hand - it's the type we expect, not the format. But it's a perfectly legal call.  In 1.15.0 (specifically #5112) the signature was changed to the following:
```py3
def load_model(
    f: IO[bytes] | str | os.PathLike,
    format: _SupportedFormat | None = None,
    load_external_data: bool = True,
) -> ModelProto:
```
Which is a narrower definition than the original, and it now rejects previously valid code because `onnx.ModelProto` is the format of the data, but not a known serialization format. The error message specifically is this:
```py3
  ValueError: Unsupported format: '<class 'onnx.onnx_ml_pb2.ModelProto'>'. Supported formats are: dict_keys(['protobuf', 'textproto', 'json', 'onnxtxt'])
```

### Expected behavior

Breaking changes do not occur in minor versions ;-)

**Comment by gramalingam** [comment]
Good point! Worth discussing in the ONNX meetings. ONNX is more careful about compatibility with respect to opset versioning, and IR versioning, and less strict about following semver with respect to release-versioning (as Justin says). However, it does try preserve compatibility where possible (but bug-fixes are expected and allowed). Another related issue that ONNX needs to address is have a clear demarcation between internal APIs and external APIs (where backward-compatibility is not guaranteed for internal APIs), which is sadly missing currently.

**Comment by cbourjau** [comment]
Did `model_proto = onnx.load_model(f, onnx.ModelProto)` actually do anything or did it just work because it is the default anyhow? If it actually just didn't have an effect, I'd classify this as a bug fix which is also reflected in a better type hint. Bug fixes are usually not semver breaking changes.

**Comment by tgolsson** [comment]
The value wasn't used, and the type hint said anything would work. So the prescribed behavior and the observed behavior matched, and the call was correct by the API. I'd say if the original API had said the only legal value was *None*, I'd agree with you. But we followed the API, and then the API was changed to have a narrower set of values without deprecation. 

I mostly filed this for someone else to find it while searching, and because it's funny that even an ineffectual argument could break in a future API. :D
